### PR TITLE
[ENH] MNE Analyze loading bar, bug fixes

### PR DESCRIPTION
**Pull request by gabrielbmotta**
- Add loading bar widget to status bar.
- Add loading stack to status bar.
- Added hover widget to display current loading processes
- Moved detect trigger to separate thread to prevent freeze/stutter.
- Detecting events from triggers now triggers loading bar to appear during calculation.
- Fix behavior of no user channel selection not displaying all channels

### CODE REVIEW

File: applications/mne_analyze/plugins/annotationmanager/annotationsettingsview.h [hunk 1]
```diff
@@ -360,15 +372,26 @@ private slots:
...
+    /**
+     * Perfroms trigger detection and sorts events into map of events by group based on detection threshold
+     *
+     * @param[in] sChannelName      name of stim channel from which we will be reading
+     * @param[in] dThreshold        threshold for a spike to count as a trigger
+     *
+     * @return      returns map of events sorted by groups based on threshold
+     */
+    QMap<double,QList<int>> detectTriggerCalculations(const QString& sChannelName,
+                                   double dThreshold);
```
Comment (LorenzE) [inline applications/mne_analyze/plugins/annotationmanager/annotationsettingsview.h:393]: mind indentation

File: libraries/disp/viewers/triggerdetectionview.h [hunk 2]
```diff
@@ -121,6 +121,14 @@ class DISPSHARED_EXPORT TriggerDetectionView : public AbstractView
     void setNumberDetectedTriggersAndTypes(int totalNumberDetections,
                                            const QMap<int,QList<QPair<int,double> > >& mapDetectedTriggers);
 
+    //=========================================================================================================
+    /**
+     * Gets currently selected STIM channel in GUI
+     *
+     * @return currently selected STIM channel
+     */
+    QString getChannelName();
```
Comment (LorenzE) [inline libraries/disp/viewers/triggerdetectionview.h:130]: rename to `getSelectedStimChannel()`

**Review by LorenzE — CHANGES_REQUESTED** [review]
cool new feature! We need to think about how to handle multiple loading events send from different
plugins to the status bar. Any ideas @gabrielbmotta ?

File: applications/mne_analyze/plugins/annotationmanager/annotationsettingsview.cpp [hunk 3]
```diff
@@ -630,8 +646,10 @@ void AnnotationSettingsView::onDetectTriggers(const QString &sChannelName,
     }
 
     if(iCurrentTriggerChIndex == 9999){
-        qWarning() << "[AnnotationSettingsView::onDetectTriggers] Channel Index not valid";
-        return;
+        qWarning() << "[AnnotationSettingsView::onDetectTriggers] Channel Index not valid";\
+        emit loadingEnd();
```
[the remark this answers is not in the document]
Comment (gabrielbmotta) [reply applications/mne_analyze/plugins/annotationmanager/annotationsettingsview.cpp:650]: This is leftover from my attempt to implement this without QFuture. Will remove.

File: applications/mne_analyze/plugins/annotationmanager/annotationsettingsview.cpp [hunk 4]
```diff
@@ -650,31 +668,15 @@ void AnnotationSettingsView::onDetectTriggers(const QString &sChannelName,
 
     for(QPair<int,double> pair : detectedTriggerSamples){
         mEventsinTypes[pair.second].append(pair.first);
+        QApplication::processEvents();
```
[the remark this answers is not in the document]
Comment (gabrielbmotta) [reply applications/mne_analyze/plugins/annotationmanager/annotationsettingsview.cpp:671]: This is leftover from my attempt to implement this without QFuture. Will remove.

**Comment by gabrielbmotta** [comment]
With multiple loads we could maybe have a loading stack where each new loadingStart adds an item to it, and only the most recent loading message gets diplayed, then while whichever one is the most recently received one is at the top, and items get popped of if they are done, with an item in the stack being marked done if they have received a corresponding loadingEnd.

**Comment by LorenzE** [comment]
> With multiple loads we could maybe have a loading stack where each new loadingStart adds an item to it, and only the most recent loading message gets diplayed, then while whichever one is the most recently received one is at the top, and items get popped of if they are done, with an item in the stack being marked done if they have received a corresponding loadingEnd.

Or maybe we can have multiple loading bars displayed for every process currently running? Something like a drop down widget which shows all loading bars when selected.

**Comment by LorenzE** [comment]
The name of the generated average model should include the name of the events group.

File: applications/mne_analyze/libs/anShared/Management/statusbar.cpp [hunk 5]
```diff
@@ -97,10 +112,83 @@ void StatusBar::onNewMessageReceived(const QSharedPointer<Event> pEvent)
...
+        }
         default:
             qWarning() << "[StatusBar::onNewMessageReceived] Received a message/event that is not handled by switch-cases";
             break;
     }
 }
 
 //=============================================================================================================
+
+void StatusBar::enterEvent(QEvent *event){
```
Comment (LorenzE) [inline applications/mne_analyze/libs/anShared/Management/statusbar.cpp:157]: linebreak before {

File: applications/mne_analyze/plugins/annotationmanager/annotationmanager.h [hunk 6]
```diff
@@ -147,6 +147,20 @@ class ANNOTATIONMANAGERSHARED_EXPORT AnnotationManager : public ANSHAREDLIB::IPl
      */
     void onJumpToSelected();
 
+    //=============================================================================================================
+    /**
+     * Sends event to trigger loading bar to appear and sMessage to show
+     *
+     * @param [in] sMessage     loading bar message
+     */
+    void triggerLoadingStart(QString sMessage);
```
Comment (LorenzE) [inline applications/mne_analyze/plugins/annotationmanager/annotationmanager.h:156]: use `const QString&`

File: applications/mne_analyze/plugins/annotationmanager/annotationmanager.h [hunk 7]
```diff
@@ -147,6 +147,20 @@ class ANNOTATIONMANAGERSHARED_EXPORT AnnotationManager : public ANSHAREDLIB::IPl
...
+
+    //=============================================================================================================
+    /**
+     * Sends event to hide loading bar
+     */
+    void triggerLoadingEnd(QString sMessage);
+
```
Comment (LorenzE) [inline applications/mne_analyze/plugins/annotationmanager/annotationmanager.h:163]: use `const QString&`

**Review by LorenzE — CHANGES_REQUESTED** [review]

**Comment by gabrielbmotta** [comment]
> The name of the generated average model should include the name of the events group.

events
Something like this?

**Comment by LorenzE** [comment]
another thing: The detect from trigger channel for sample_audvis_raw.fif and trigger channel STI014 on MacOS did not produce any event groups (which it did before).

**Comment by gabrielbmotta** [comment]
> another thing: The detect from trigger channel for sample_audvis_raw.fif and trigger channel STI014 on MacOS did not produce any event groups (which it did before).

Are the other channels working?

**Comment by LorenzE** [comment]
> >

sry was on the wrong branch. works fine!

**Review by LorenzE — DISMISSED** [review]

**Review by LorenzE — APPROVED** [review]

**Pull request merged**